Datapath work for adding bpf metrics map

manalibhutiyani opened the pull request:
This PR adds the ground work needed to support prometheus L3/L4 packet drops and forwards.
This adds the data path bits in place. 

Fixes: #4155
Related-to: #3284
Related-to: #3337

```release-note
Datapath work for adding bpf L3-L4 drop/forward metrics map
```

Commit:
cmd: Add a CLI command to access the bpf L3-L4 metrics map

This change adds a new `cilium bpf metrics list` command to access the
new metrics bpf map.

Fixes: #4155
Related-to: #3284
Related-to: #3337
Signed-Off-By: Manali Bhutiyani <manali@covalent.io>

Commit:
bpf: Add BPF map cilium_metrics for L3-l4 packet drops/forwards

This commit adds a new global BPF map named cilium_metrics which keeps account for
L3-L4 packet drops and forwards. The key of the map is the drop reason and direction (ingress/egress),
while the value is the count. If drop reason = 0 then the packet is forwarded and the count is the
count of forwarded packets.

Fixes: #4155
Related-to: #3284
Related-to: #3337
Signed-Off-By: Manali Bhutiyani <manali@covalent.io>

Commit:
pkg/maps/metricsmap: Add a new userspace pkg/maps/metricsmap to access BPF metrics maps.

This commit adds a new pkg/maps/metricsmap which defines the userspace interface and
functions to access the BPF metrics maps.

Fixes: #4155
Related-to: #3284
Related-to: #3337
Signed-Off-By: Manali Bhutiyani <manali@covalent.io>

tgraf commented on bpf/bpf_lxc.c:
```diff
@@ -691,10 +693,13 @@ int handle_ingress(struct __sk_buff *skb)
-	if (IS_ERR(ret))
-		return send_drop_notify(skb, SECLABEL, 0, 0, 0, ret, TC_ACT_SHOT);
-	else
+	if (IS_ERR(ret)) { // dropped
```
Please use proper C comments `/* */`

manalibhutiyani replied:
will do

manalibhutiyani replied:
done

tgraf commented on bpf/bpf_lxc.c:
```diff
@@ -691,10 +693,13 @@ int handle_ingress(struct __sk_buff *skb)
+		return send_drop_notify(skb, SECLABEL, 0, 0, 0, ret, TC_ACT_SHOT,
+		                        DIRECTION_EGRESS);
+	} else { //forwarded
+	    update_data_metrics(DIRECTION_EGRESS, 0);
```
This is incorrect, there are various forwarding paths where we tail call and never reach this pointed even when forwarding. Look at the invocations of `DBG_CAPTURE_DELIVERY` to see all places where we forward a packet.

manalibhutiyani replied:
Sounds good, thanks for the pointer.

manalibhutiyani replied:
@tgraf I am planning to add the forwarded counters before all invocations of `cilium_dbg_capture(skb, DBG_CAPTURE_DELIVERY, ifindex);` . This includes bpf_lb.c, icmp6.h, arp.h, bpf_lxc.c and bpf_overlay.c,  since the dropped counters are updated on all invocations of `send_drop_notify` independent of the whether the packet is any of the following: IPV4/IPV6 ICMP/TCP/UDP/ARP. Please let me know if that's correct?

manalibhutiyani replied:
@tgraf :Also  `DBG_CAPTURE_DELIVERY` is applicable for `ingress` as well as `egress`?
So the forwarded count code added in `handle_policy` is incorrect for ingress?

manalibhutiyani replied:
@joestringer @tgraf : What will the direction be for`cilium_dbg_capture(skb, DBG_CAPTURE_DELIVERY, skb->ifindex)`  in  `icmp6_send_reply`. and `arp_respond`. I think it should be EGRESS.
Similarly, going through `bpf_overlay.c`, `cilium_dbg_capture(skb, DBG_CAPTURE_DELIVERY, skb->ifindex);` in this case both should be egress, while in `bpf_lb.c` its ingress correct?

manalibhutiyani replied:
As discussed with @joestringer , since packet can be forwarded at multiple places, incrementing at say both netdev and lxc might lead to the same packet being incremented twice. Hence I am conforming the forward counts mainly to bpf_lxc.c right now. Please let me know if that's fine.

manalibhutiyani replied:
Spoke to @rlenglet. We can use `send_trace_notify` instead to specifically capture forwarded packets.

manalibhutiyani replied:
done

tgraf commented on bpf/bpf_lxc.c:
```diff
@@ -961,9 +966,14 @@ __section_tail(CILIUM_MAP_POLICY, LXC_ID) int handle_policy(struct __sk_buff *sk
 		return send_drop_notify(skb, src_label, SECLABEL, LXC_ID,
-					ifindex, ret, TC_ACT_SHOT);
+					ifindex, ret, TC_ACT_SHOT, DIRECTION_INGRESS);
+	} else { // forwarded
```
The else is not required as you return above

manalibhutiyani replied:
by bad. will fix it

manalibhutiyani replied:
done

tgraf commented on bpf/lib/common.h:
```diff
@@ -188,6 +188,19 @@ struct policy_entry {
 
+struct data_metrics_key {
+    __u8      reason;     //0: forwarded, >0 dropped
+    __u8      dir:1,      //0: Unknown 1: ingress 2: egress
```
This doesn't make sense, `2` uses 2 bits. When is the direction unknown? We can just use a single bit for ingress/egress.

manalibhutiyani replied:
Sorry. this was initially 0 and 1 and then I changed this to 0, 1,2 and did not change the number of  bits. 
The 0 or unknown is for when we might not know the ingress/egress direction in the code. Let me change that

manalibhutiyani replied:
Also, do you think 0 or unknown direction is not needed?

manalibhutiyani replied:
@tgraf `send_drop_notify_error` calls `send_drop_notify` and we do not always know the direction in that case. Hence the `UNKNOWN`. Let me know your thoughts on this

manalibhutiyani replied:
Changing the number of bits to 2. Please let me know if you feel otherwise.

joestringer replied:
I think it's important to define what this "ingress" / "egress" means in a comment. I assume that "ingress" here is describing traffic traversing to an endpoint, and egress is traversing from an endpoint.

"Unknown" to me suggests that the traffic is received to the node, but we don't intend to forward it to an endpoint inside the node. I believe that in all other cases, the traffic should either be clearly being sent by an endpoint, or being sent towards an endpoint. 

That said, I'm not sure whether there are currently any cases inside the BPF code where we are receiving traffic from outside the node and handle forwarding to another node. AFAIK such cases are handled instead by regular linux stack constructs today, such as the route table, etc.

manalibhutiyani replied:
@joestringer `send_drop_notify_error` is called from a variety of other places, bpf_lb.c, bpf_overlay.c and right now we fill the direction to be 0 

See
```
static inline int send_drop_notify_error(struct __sk_buff *skb, int error, int exitcode)
{
	return send_drop_notify(skb, 0, 0, 0, 0, error, exitcode, 0);
}
```
This is what I mean by "unknown".  Let me know if this is correct.

manalibhutiyani replied:
done

tgraf commented on bpf/lib/common.h:
```diff
@@ -188,6 +188,19 @@ struct policy_entry {
+
+
+struct data_metrics_value {
+     __u64		count;
```
We also want a bytes counter here

manalibhutiyani replied:
sure

manalibhutiyani replied:
done

tgraf commented on daemon/daemon.go:
```diff
@@ -803,6 +804,7 @@ func (d *Daemon) init() error {
 	fmt.Fprintf(fw, "#define TUNNEL_ENDPOINT_MAP_SIZE %d\n", tunnel.MaxEntries)
 	fmt.Fprintf(fw, "#define PROXY_MAP_SIZE %d\n", proxymap.MaxEntries)
 	fmt.Fprintf(fw, "#define ENDPOINTS_MAP_SIZE %d\n", lxcmap.MaxEntries)
+	fmt.Fprintf(fw, "#define METRICS_MAP_SIZE %d\n", datametrics.MaxEntries)
```
This commit needs to be merged with the previous commit as it breaks bisecting otherwise.

manalibhutiyani replied:
will do

manalibhutiyani replied:
done

Commit:
linux/bpf.h: Add reference link to in-kernel sk_buff structure.

This is particularly useful to map the bpf sk_buff which should always be in sync with
the  in-kernel sk_buff structure.

Fixes: #4155
Related-to: #3284
Related-to: #3337
Signed-Off-By: Manali Bhutiyani <manali@covalent.io>

jrfastab commented on bpf/lib/common.h:
```diff
@@ -188,6 +188,20 @@ struct policy_entry {
 };
 
+struct data_metrics_key {
+    __u8      reason;     //0: forwarded, >0 dropped
```
There should probably be an enum for the 'reason' as well, METRIC_REASON_FORWARDED, METRIC_REASON_DROPPED, ... to support more metrics in the future.

Also just a question, would it be easier to encode the direction in  the reason rather than have a dir bit? So
enum {
  METRIC_REASON_FORWARD_INGRESS,
  METRIC_REASON_FORWARD_EGRESS,
  METRIC_REASON_DROP_INGRESS,
  METRIC_REASON_DROP_EGRESS,
};

In the future we could extend this to be even more specific if needed, METRIC_REASON_DROP_SOCKET for example comes to mind.  I'm not entirely up to speed on the prometheous reqs so maybe the bit has some meaning for mapping into higher layers?

manalibhutiyani replied:
The reason can be any of the above combinations:
If reason is 0 it is forwarded and the value is interpreted as forwarded count, else
reason is the drop reason as below:
```
/* Cilium error codes, must NOT overlap with TC return codes */
#define DROP_INVALID_SMAC	-130
#define DROP_INVALID_DMAC	-131
#define DROP_INVALID_SIP	-132
#define DROP_POLICY		-133
#define DROP_INVALID		-134
#define DROP_CT_INVALID_HDR	-135
#define DROP_CT_MISSING_ACK	-136
#define DROP_CT_UNKNOWN_PROTO	-137
#define DROP_CT_CANT_CREATE	-138 /* unused */
#define DROP_UNKNOWN_L3		-139
#define DROP_MISSED_TAIL_CALL	-140
#define DROP_WRITE_ERROR	-141
#define DROP_UNKNOWN_L4		-142
#define DROP_UNKNOWN_ICMP_CODE	-143
#define DROP_UNKNOWN_ICMP_TYPE	-144
#define DROP_UNKNOWN_ICMP6_CODE	-145
#define DROP_UNKNOWN_ICMP6_TYPE	-146
#define DROP_NO_TUNNEL_KEY	-147
#define DROP_NO_TUNNEL_OPT	-148 /* unused */
#define DROP_INVALID_GENEVE	-149 /* unused */
#define DROP_UNKNOWN_TARGET	-150
#define DROP_NON_LOCAL		-151
#define DROP_NO_LXC		-152
#define DROP_CSUM_L3		-153
#define DROP_CSUM_L4		-154
#define DROP_CT_CREATE_FAILED	-155
#define DROP_INVALID_EXTHDR	-156
#define DROP_FRAG_NOSUPPORT	-157
#define DROP_NO_SERVICE		-158
#define DROP_POLICY_L4		-159
#define DROP_NO_TUNNEL_ENDPOINT -160
#define DROP_PROXYMAP_CREATE_FAILED	-161
#define DROP_POLICY_CIDR		-162
```
Extending this to say `REASON_DROP_SOCKET` will just require adding a new `DROP_REASON`. 
Also making a combination of these DROP_REASON with ingress/egress integrated does not seem to add value IMO (there will be around 62 combinations total), but let me know if there are any performance disadvantages, of doing so? Hence I decided to keep the direction separate instead of coupling it with reason.

jrfastab replied:
OK seems reasonable, thanks for the explanation. The '-reason' is a bit ugly but I don't see any good way to fix that easily and its a really big nit on my part.

jrfastab commented on bpf/lib/common.h:
```diff
@@ -247,6 +261,8 @@ enum {
 #define DROP_PROXYMAP_CREATE_FAILED	-161
 #define DROP_POLICY_CIDR		-162
 
+#define DIRECTION_INGRESS 0
```
I think an enum would be better then #define each direction.

manalibhutiyani replied:
sure.

manalibhutiyani replied:
done

jrfastab commented on bpf/lib/l3.h:
```diff
@@ -88,6 +88,13 @@ static inline int ipv6_local_delivery(struct __sk_buff *skb, int l3_off, int l4_
+	 * Note that the packet could still be dropped but it would show up
+	 * as an ingress drop counter in metrics.
+	 */
+	update_data_metrics(skb->len, DIRECTION_EGRESS, 0);
```
Here is where the enum reason makes things more readable vs '0'

jrfastab replied:
Is there any value in recording this special case as its own type LXC_FORWARD for example?

manalibhutiyani replied:
0 is for forwarded. Not the direction. I can add a separate #define just for forwarded reason, to make it more readable.

Regarding `LXC_FORWARD` , I don't think we need that level of granularity yet from prometheus perspective, and hence I am using the first step for exposing total drops/forwards in each direction.


In prometheus, I envision to display total drops/drop reason and total drops/direction.
Similarly for forwards. I can make a note of this and do it as a separate PR if we decide to add more granularity going forward.

Let me know.

jrfastab replied:
Sounds good, we can add more granularity when/if its needed. No need to do it now. Maybe just a #define for FORWARD and I'm happy. Again just nit-picking.

manalibhutiyani replied:
done

jrfastab commented on bpf/lib/metrics.h:
```diff
@@ -0,0 +1,62 @@
+    struct data_metrics_key key = {};
+
+    key.reason = reason;
+    key.dir     = direction;
```
are those '=' aligned or is my browser not rendering correctly.

manalibhutiyani replied:
My bad. I will fix it. This is what happens when you use golang editor for C code :)

manalibhutiyani replied:
done

jrfastab commented on bpf/lib/metrics.h:
```diff
@@ -0,0 +1,62 @@
+
+    if ((entry = map_lookup_elem(&cilium_metrics, &key))) {
+            __sync_fetch_and_add(&entry->count, 1);
+            __sync_fetch_and_add(&entry->bytes, bytes);
```
Is there any performance impact from using sync_fetch_and_add here? How about using a per cpu map and aggregating when we pull the metrics out into golang side.

manalibhutiyani replied:
Yes, we need to make this a per CPU map, and I will be raising a separate GH issue for that.
I am starting this groundwork off by the HASH_MAP to get it working end-to-end for 1.1

jrfastab replied:
Fair enough can you create the GH issue and link to it from the PR for this work just so we don't lose it.

manalibhutiyani replied:
will do 👍

manalibhutiyani replied:
Raised #4196 . Will put in in comments too

manalibhutiyani replied:
done

joestringer commented on cilium/cmd/bpf_metrics_list.go:
```diff
@@ -0,0 +1,61 @@
+)
+
+var bpfMetricsListCmd = &cobra.Command{
+	Use:   "metrics list",
```
Please create a separate file, `cilium/cmd/bpf_metrics.go` which just adds a subcommand, then change this to "list". Compare against eg the recent `cilium bpf ipcache ...` changes. Otherwise the help messages and so on don't work correctly.

manalibhutiyani replied:
will do

manalibhutiyani replied:
done

joestringer commented on daemon/daemon.go:
```diff
@@ -69,6 +69,7 @@ import (
 	"github.com/cilium/cilium/pkg/workloads"
 	"github.com/cilium/cilium/pkg/workloads/containerd"
 
+	"github.com/cilium/cilium/pkg/maps/datametrics"
```
Real minor nit, but why "datametrics", and not just "metrics"? It's already clear that it's for the datapath because it's a map (or the underlying code is in the datapath).

manalibhutiyani replied:
Because I wanted to make this a package and we already have `metrics` package in `pkg/metrics`. I initially thought that since this is under `maps`, defining  a package named `metrics` in `pkg/maps ` will still work but I get errors in that case, when I try to use the `pkg/metrics` somewhere. Hence I named it to `datametrics` to keep things simple. Please let me know if there is a golang style I need to be aware of here?

manalibhutiyani replied:
I did not want to combine the `pkg/metrics` with this one, because `pkg/metrics` is very specific to prometheus, while this was more relevant to L3-L4 data path.

joestringer replied:
From the golang side, most of the maps seem to be (admittedly, also redundantly...) called xxxmap, so following the style would call this metricsmap.

From the BPF datapath code, I don't think it's necessary to differentiate as this is the only metrics being performed there.

manalibhutiyani replied:
Changed `datametrics` --> `metricsmap` on golang side.
And on BPF side, changed `data_metrics` to only `metrics`.

joestringer commented on pkg/maps/datametrics/datametrics.go:
```diff
@@ -0,0 +1,93 @@
+
+// NewValue returns a new empty instance of the structure representing the BPF
+// map value
+func (k Key) NewValue() bpf.MapValue { return &Value{} }
```
Please use pointer receivers here. We hit issues some time ago with non-pointer receivers in map keys.

(Looks like ipcache needs the same fix, filed https://github.com/cilium/cilium/issues/4192)

manalibhutiyani replied:
will do

manalibhutiyani replied:
I am not really using this functionality, so removing it altogether.

manalibhutiyani replied:
I checked and this is the definition of `//NewValue creates a new bpf.MapValue.
func (k *CtKey6) NewValue() bpf.MapValue { return &CtEntry{} }` in all other places... 
I think this function needs to be declared with this same syntax to properly identify the `key` interface? Please search `NewValue` .

joestringer replied:
Yes, all of the functions with the Key receiver will need to be pointers. ie

`func (k *Key) NewValue() bpf.MapValue { return &Value{} }`,
not
`func (k Key) NewValue() bpf.MapValue { return &Value{} }`

joestringer commented on bpf/bpf_overlay.c:
```diff
@@ -141,7 +141,7 @@ __section_tail(CILIUM_MAP_CALLS, CILIUM_CALL_IPV4) int tail_handle_ipv4(struct _
 
 	if (IS_ERR(ret))
-		return send_drop_notify_error(skb, ret, TC_ACT_SHOT);
+		return send_drop_notify_error(skb, ret, TC_ACT_SHOT, DIRECTION_EGRESS); //Check if direction is correct
```
The section includes `CILIUM_CALL_IPV4`. Tracing the `ep_tail_call(...)` call for this enum shows that it is in the section `from-overlay`. This suggests that it should actually be `DIRECTION_INGRESS`. Furthermore, tracing that function to `handle_ipv6` inside `bpf/bpf_overlay.c` shows calls to `skb_get_tunnel_key()`, ie it's fetching info from the tunnel metadata, and also `ipv6_local_delivery()`, so it's sending to the local stack (ie ingress), not into the tunnel (ie egress). 

Similarly for the below call.

manalibhutiyani replied:
thanks.. will change

manalibhutiyani replied:
done

joestringer commented on bpf/lib/icmp6.h:
```diff
@@ -121,7 +121,7 @@ __section_tail(CILIUM_MAP_CALLS, CILIUM_CALL_SEND_ICMP6_ECHO_REPLY) int tail_icm
 	ret = __icmp6_send_echo_reply(skb, nh_off);
 	if (IS_ERR(ret))
-		return send_drop_notify_error(skb, ret, TC_ACT_SHOT);
+		return send_drop_notify_error(skb, ret, TC_ACT_SHOT, DIRECTION_EGRESS); //Check if direction is correct
```
For this one, you can trace similar to the overlay case by checking where `CILIUM_CALL_SEND_ICMP6_ECHO_REPLY` is used.

manalibhutiyani replied:
It's called in `icmp6_handle` for case `ICMPV6_ECHO_REQUEST` . So I am assuming this should be `DIRECTION_INGRESS` too

manalibhutiyani replied:
done. Please check

joestringer reviewed (dismissed):
Some minor comments, I didn't look too deeply into the BPF since it looked like John had provided several comments already.

tgraf commented on bpf/lib/common.h:
```diff
@@ -188,6 +188,20 @@ struct policy_entry {
+
+struct metrics_value {
+     __u64	count;
+     __u64      bytes;
```
Use tabs properly here

manalibhutiyani replied:
will do

manalibhutiyani replied:
done

tgraf commented on bpf/lib/common.h:
```diff
@@ -247,6 +261,14 @@ enum {
 #define DROP_POLICY_CIDR		-162
 
+/* Cilium metrics reason for forwarding packet */
+#define REASON_FORWARDED  0
```
Add a comment which states that reason > 0 corresponds to -(DROP_*)

manalibhutiyani replied:
sure

manalibhutiyani replied:
done

tgraf commented on bpf/lib/icmp6.h:
```diff
@@ -303,7 +303,7 @@ __section_tail(CILIUM_MAP_CALLS, CILIUM_CALL_SEND_ICMP6_TIME_EXCEEDED) int tail_
 	ret = __icmp6_send_time_exceeded(skb, nh_off);
 	if (IS_ERR(ret))
-		return send_drop_notify_error(skb, ret, TC_ACT_SHOT);
+		return send_drop_notify_error(skb, ret, TC_ACT_SHOT, DIRECTION_EGRESS);
```
This can be both ingress and egress. The direction must be passed in from the original callers.

manalibhutiyani replied:
done

tgraf commented on bpf/lib/icmp6.h:
```diff
@@ -121,7 +121,7 @@ __section_tail(CILIUM_MAP_CALLS, CILIUM_CALL_SEND_ICMP6_ECHO_REPLY) int tail_icm
 	ret = __icmp6_send_echo_reply(skb, nh_off);
 	if (IS_ERR(ret))
-		return send_drop_notify_error(skb, ret, TC_ACT_SHOT);
+		return send_drop_notify_error(skb, ret, TC_ACT_SHOT, DIRECTION_INGRESS);
```
This can be both ingress and egress. The direction must be passed in from the original callers.

manalibhutiyani replied:
I believe this is being called by the tail call `CILIUM_CALL_SEND_ICMP6_ECHO_REPLY` which is only called at `icmp6_send_echo_reply`. I guess we are replying back to case `ICMPV6_ECHO_REQUEST` in `icmp6_handle`. So since we are replying ourselves to the ICMP reply, I can see that this case is `EGRESS`? Which is the `INGRESS` scenario? @joestringer @tgraf

manalibhutiyani replied:
done

tgraf commented on bpf/lib/icmp6.h:
```diff
@@ -355,7 +355,7 @@ __section_tail(CILIUM_MAP_CALLS, CILIUM_CALL_HANDLE_ICMP6_NS) int tail_icmp6_han
 	ret = __icmp6_handle_ns(skb, nh_off);
 	if (IS_ERR(ret))
-		return send_drop_notify_error(skb, ret, TC_ACT_SHOT);
+		return send_drop_notify_error(skb, ret, TC_ACT_SHOT, DIRECTION_EGRESS);
```
This can be both ingress and egress. The direction must be passed in from the original callers.

manalibhutiyani replied:
done

tgraf commented on bpf/lib/trace.h:
```diff
@@ -100,6 +101,20 @@ static inline void send_trace_notify(struct __sk_buff *skb, __u8 obs_point, __u3
+		case TRACE_FROM_HOST:
+		case TRACE_FROM_STACK:
+		case TRACE_FROM_OVERLAY:
+			update_metrics(skb->len, DIRECTION_EGRESS, REASON_FORWARDED);
```
This is incorrect. TRACE_FROM_ is no indication for a successful forwarding. It can still be dropped afterwards. We may only increment REASON_FORWARD for DIRECTION_EGRESS on successful delivery.

You need to look at the call sites of `DBG_CAPTURE_DELIVERY` and increment it there.

manalibhutiyani replied:
I think I can still use `send_trace_notify` but the logic should change like so:

```
static inline void send_trace_notify(struct __sk_buff *skb, __u8 obs_point, __u32 src, __u32 dst,
				     __u16 dst_id, __u32 ifindex, __u8 reason)
{
	switch (obs_point) {
		case TRACE_TO_LXC:
			update_metrics(skb->len, DIRECTION_INGRESS, REASON_FORWARDED);
			break;

		/* TRACE_FROM_LXC is handled separately in ipv*_local_delivery() */
		case TRACE_TO_HOST:
		case TRACE_TO_STACK:
		case TRACE_TO_OVERLAY:
			update_metrics(skb->len, DIRECTION_EGRESS, REASON_FORWARDED);
	}
}
```


This way `TRACE_TO_LXC` is specifically forwarded to endpoint successfully at `ingress` and rest all `TRACE_TO*` (leaving out the PROXY cases) will be forwarding at egress.

We need to handle the special case of endpoint to endpoint i.e `egress` forwarding at `ipv*_local_delivery()` 


Is that correct? `DBG_CAPTURE_DELIVERY` is selectively called along with `send_trace_notify` when we are only concerned with endpoints i.e which is also the scope of this PR.


Please let me know if that's the correct logic, i.e we can simply change `send_trace_notify` and not `cilium_dbg_capture` since `send_trace_notify` is a subset of all the meaningful datapoints we need from cilium_dbg_capture/`DBG_CAPTURE_DELIVERY`

If this is still incorrect, I would like to schedule a call to discuss this further to avoid further delay on my part in understanding this.

manalibhutiyani replied:
done

aanm commented on pkg/maps/metricsmap/metricsmap.go:
```diff
@@ -0,0 +1,93 @@
+}
+
+// GetKeyPtr returns the unsafe pointer to the BPF key
+func (k Key) GetKeyPtr() unsafe.Pointer { return unsafe.Pointer(&k) }
```
Please change it to `func (k *Key) GetKeyPtr() unsafe.Pointer { return unsafe.Pointer(k) }`

manalibhutiyani replied:
will do

manalibhutiyani replied:
done
